### ISSUE TITLE
LSP ranges are incorrect when line contains multi-codepoint grapheme clusters

### ISSUE THREAD

**Issue by yoshi-monster** [gleam-lang/gleam#3628]
The string is `"क्षि"`, or `"\u{0915}\u{094D}\u{0937}\u{093F}"`.

![image](https://github.com/user-attachments/assets/6ee7eee7-dd81-45ca-94c9-37c0dfc1999a)

From the looks of it it actually just uses the byte/code unit count.

Tested with Gleam v1.4.1, in VSCode, Zed and Helix :purple_heart:

**Comment by lpil** [comment]
Thank you. I expect this'll be due to LSP using utf-16 and we're using utf-8.

**Comment by aaron-loa** [comment]
Hello. I looked a bit into this. 
Supposedly by setting the position_encoding type to UTF-8 the problem should be solved.
After specifying the encoding it works correctly with editors that support UTF-8 encoding, like neovim. 
But it doesn't work with vscode.
This is the error I get:
`Error: Unsupported position encoding (utf-8) received from server Gleam Language Server`

Not sure about zed, I can't figure out how to change the path of the gleam language server to use my local binary.

Hovering is also affected by this problem, probably everything that interacts with positions.
I will try to look more into what can be done this week. But from a quick skimming through, a lot of things need to be changed :smile:

**Comment by lpil** [comment]
Thank you @aaron-loa !!

**Comment by lydell** [comment]
> probably everything that interacts with positions

Yep, that seems to be the case. Here’s an example with the “Add missing patterns“ code action:

```gleam
// ä
fn wibble() {
  case True {
    
  }
}
```

Becomes:

```gleam
// ä
fn wibble() {
  case True 
   False -> todo
   True -> todo
 }
}
```

**Comment by aaron-loa** [comment]
I did some investigation, but completely forgot to write an update.

In compiler-core/src/line_numbers.rs the LineNumber structure doesn't do the translation between utf-8 -> to utf-16.
Ideally this should would work similary to this:
<https://github.com/artempyanykh/rust-lsp-document/blob/main/src/lib.rs>
This is a small library that does the translation between utf-16 and utf-8 indices, for language servers.

We could do this translation in the LineNumber struct, the only problem is we always need a reference to the source text. But adding the reference to the struct adds too many lifetime issues, i just gave up on this approach, but this would be the most ideal approach imo.

After this I tried a different approach, where you have to put the source text as an argument to the byte_index method, but I don't think this is a 'clean' solution, that's why I didnt' make a pr.
This is the fork where I tried it <https://github.com/aaron-loa/gleam>. I got the outgoing positions with correct UTF-16 indices, but not the information that comes from the client the language server, like hovers, couldn't figure out why. Fork is behind a lot and there are merge conflicts, but just in case someone wants to look at it I will post it here.

**Comment by GearsDatapacks** [comment]
Looking into this, this actually looks like two different issues:
- The first is when a line contains a multibyte UTF-8 character, which causes all LSP ranges to be incorrect, due to LSP using UTF-16
- The second seems to only occur in certain code actions, such as add missing patterns, where a multibyte character anywhere in the source file before the cursor. I'm still trying to understand why this happens.
